### ISSUE TITLE
Anaconda Cloud now causes error and package removal for upload of package version of same name

### ISSUE THREAD

**Issue by matthewfeickert** [Anaconda-Platform/anaconda-client#702]
Copying https://github.com/scientific-python/upload-nightly-action/issues/57 here to get feedback from the Anaconda team. This is a breaking behavior change, so advice on how to proceed would be quite welcome.

> Anaconda Cloud seems to have changed how they handle uploads of packages where there is already a package of the same name and version on Anaconda Cloud. In the past, this simply just resulted in the artifact being uploaded overwriting the previous artifact. `networkx` is one of the core packages that used this approach. However, I noticed that [`networkx`'s nightly workflow](https://github.com/networkx/networkx/blob/d6569adf7c224a9c85c0e4e4c92435f772f73582/.github/workflows/nightly.yml) is failing to do so
>
> ![image](https://github.com/scientific-python/upload-nightly-action/assets/5142394/2c7eb776-e7dc-4402-b7eb-3b363f8210e3)
>
>
> ```
> ...
> File type is "Standard Python"
> Extracting standard python attributes for upload
> Creating package "networkx"
> Creating release "3.3rc0.dev0"
> Uploading file "scientific-python-nightly-wheels/networkx/3.3rc0.dev0/networkx-3.3rc0.dev0-py3-none-any.whl"
> Warning:  Distribution "networkx-3.3rc0.dev0-py3-none-any.whl" already exists. Removing.
> Error:  ("release version='3.3rc0.dev0' does not exist", 404)
> ##[debug]Docker Action run completed with exit code 1
> ##[debug]Finishing: Upload nighlty wheel
> ```
>
> (The success after failure is due to the package being removed and so there is no conflict the next upload.)
>
> This started happening on 2024-01-04 for `networkx`, and was not happening before that. For example, this 2024-01-02 upload ran successfully
> 
> ```
> + anaconda --token *** upload --force --user scientific-python-nightly-wheels dist/networkx-3.3rc0.dev0-py3-none-any.whl
> ...
> Using Anaconda API: https://api.anaconda.org/
> Using "scientific-python-nightly-wheels" as upload username
> Processing "dist/networkx-3.3rc0.dev0-py3-none-any.whl"
> Detecting file type...
> File type is "Standard Python"
> Extracting standard python attributes for upload
> Creating package "networkx"
> Creating release "3.3rc0.dev0"
> Uploading file "scientific-python-nightly-wheels/networkx/3.3rc0.dev0/networkx-3.3rc0.dev0-py3-none-any.whl"
> Warning:  Distribution "networkx-3.3rc0.dev0-py3-none-any.whl" already exists. Removing.
> 
>   0%|          | 0.00/1.59M [00:00<?, ?B/s]
> 1.59MB [00:00, 5.26MB/s]                   
> Upload complete
> 
> standard python located at:
>   https://anaconda.org/scientific-python-nightly-wheels/networkx
> ```
> 
> We know that as we're using a [lock file](https://github.com/scientific-python/upload-nightly-action/blob/6f1e2b99c3148480df25d296c27ae58c446ad5d0/conda-lock.yml) that nothing has changed from the `upload-nightly-action` action side, and so this has to be on the Anaconda Cloud side.

**Comment by matthewfeickert** [comment]
This behavior also seems to be invalidating the `--force` flag of `anaconda upload`

```console
$ anaconda --version
anaconda Command line client (version 1.12.1)
$ anaconda upload --help
usage: anaconda upload [-h] [-c CHANNELS] [-l LABELS] [--no-progress] [-u USER] [--keep-basename] [-p PACKAGE] [-v VERSION] [-s SUMMARY] [-t PACKAGE_TYPE] [-d DESCRIPTION] [--thumbnail THUMBNAIL]
                       [--private] [--no-register | --register] [--build-id BUILD_ID] [-i | -f | --force | --skip-existing | -m]
                       files [files ...]

Upload packages to your Anaconda repository

positional arguments:
  files                 Distributions to upload

options:
  -h, --help            show this help message and exit
  -c CHANNELS, --channel CHANNELS
                        [DEPRECATED] Add this file to a specific channel. Warning: if the file channels do not include "main", the file will not show up in your user channel
  -l LABELS, --label LABELS
                        Add this file to a specific label. Warning: if the file labels do not include "main", the file will not show up in your user label
  --no-progress         Don't show upload progress
  -u USER, --user USER  User account or Organization, defaults to the current user
  --keep-basename       Do not normalize a basename when uploading a conda package.
  --no-register         Don't create a new package namespace if it does not exist
  --register            Create a new package namespace if it does not exist
  --build-id BUILD_ID   Anaconda repository Build ID (internal only)
  -i, --interactive     Run an interactive prompt if any packages are missing
  -f, --fail            Fail if a package or release does not exist (default)
  --force               Force a package upload regardless of errors
  --skip-existing       Skip errors on package batch upload if it already exists
  -m, --force-metadata-update
                        Overwrite existing release metadata with the metadata from the package.
...
```

which is [being used in the relevant code](https://github.com/scientific-python/upload-nightly-action/blob/66bc1b6beedff9619cdff8f3361a06802c8f5874/cmd.sh#L75-L79) of https://github.com/scientific-python/upload-nightly-action/issues/57.

**Comment by matthewfeickert** [comment]
@csoja @yshmatov-anaconda @vshevchenko-anaconda Can you please comment here, or tag people on the relevant teams. This is API breaking behavior changes and so quite serious.

**Comment by matthewfeickert** [comment]
@kenodegard @dholth as you're currently reviewing PRs for the project, can you please either comment on this or notify the relevant teams?

**Comment by dholth** [comment]
We have notified the relevant teams.

**Comment by matthewfeickert** [comment]
> We have notified the relevant teams.

Thanks very much @dholth. I appreciate it.

**Comment by matthewfeickert** [comment]
So the problem gets significantly worse when you have a situation like https://github.com/scikit-hep/awkward/pull/3012#issuecomment-1935001723 where you are attenmpting to upload mutliple wheels of the same name, as each wheel can trigger this situation.

(cc @jpivarksi)

**Comment by matthewfeickert** [comment]
@dholth Ah. I think I understand what is happening now (summarized in https://github.com/scientific-python/upload-nightly-action/issues/57#issuecomment-1935365341). This might already have been obvious to you/the relevant teams.

The package directory is getting removed by the backend in mid-upload as it detects that the only file in the package has been removed. This apparently got changed in January 2024. Is it possible to have this cleanup logic know that it got triggered during an upload attempt and have it wait until _after_ the upload finishes to see if it should delete the package from the registry?

**Comment by matthewfeickert** [comment]
Closed as release `v1.12.3` fixes this (specifically PR #708). :+1:
